lib/Conversion/StructuredToMemref/StructuredToMemrefPass.cpp

```diff
@@ -0,0 +1,152 @@
+      return builder.create<UnrealizedConversionCastOp>(loc, resultType, inputs)
+          .getResult(0);
+    });
+  }
+};
+
+class LoopTypeConverter : public TypeConverter {
+public:
+  LoopTypeConverter(MLIRContext *context) {
+    // The order of type conversion is important: later ones are tried earlier.
+    addConversion([](Type type) { return type; });
+    addConversion([&](triton::PointerType ptrType) {
+      SmallVector<int64_t> strides{1};
+      auto layout =
+          StridedLayoutAttr::get(context, ShapedType::kDynamic, strides);
+
+      auto elemType = ptrType.getPointeeType();
+      auto memrefType = MemRefType::get({1}, elemType, layout);
+      return memrefType;
+    });
+  }
+};
+
+class StructuredToMemrefPass
+    : public triton::impl::StructuredToMemrefBase<StructuredToMemrefPass> {
+  using StructuredToMemrefBase<StructuredToMemrefPass>::StructuredToMemrefBase;
+
+public:
+  void getDependentDialects(DialectRegistry &registry) const override {
+    registry.insert<func::FuncDialect, arith::ArithDialect, math::MathDialect,
+                    linalg::LinalgDialect, affine::AffineDialect,
+                    scf::SCFDialect, tensor::TensorDialect,
+                    bufferization::BufferizationDialect, triton::TritonDialect,
+                    ttx::TritonTilingExtDialect, memref::MemRefDialect>();
+  }
+
+  LogicalResult convertArgsToMemrefType() {
+    auto moduleOp = getOperation();
+
+    RewritePatternSet patterns(&getContext());
+    ConversionTarget target(getContext());
+    TritonFunctionSignatureConverter typeConverter;
+
+    // Update function signature to use memrefs
+    target.addDynamicallyLegalOp<func::FuncOp>([&](func::FuncOp op) {
+      return typeConverter.isSignatureLegal(op.getFunctionType());
+    });
+
+    populateFunctionOpInterfaceTypeConversionPattern<func::FuncOp>(
+        patterns, typeConverter);
+
+    return applyPartialConversion(moduleOp, target, std::move(patterns));
+  }
+
+  void runOnOperation() override {
+
+    if (failed(convertArgsToMemrefType())) {
+      signalPassFailure();
+    }
+
+    auto moduleOp = getOperation();
+
+    RewritePatternSet patterns(&getContext());
+    ConversionTarget target(getContext());
+
+    target.addLegalDialect<
+        func::FuncDialect, arith::ArithDialect, math::MathDialect,
+        linalg::LinalgDialect, affine::AffineDialect, scf::SCFDialect,
+        cf::ControlFlowDialect, tensor::TensorDialect,
+        bufferization::BufferizationDialect, ttx::TritonTilingExtDialect,
+        memref::MemRefDialect>();
+
+    target.addIllegalDialect<tts::TritonStructuredDialect>();
+
+    target.addDynamicallyLegalOp<UnrealizedConversionCastOp>([](Operation *op) {
+      auto resType = op->getResultTypes()[0];
+      return !isa<triton::PointerType>(resType);
+    });
+
+    LoopTypeConverter loopTypeConverter(patterns.getContext());
+    mlir::scf::populateSCFStructuralTypeConversionsAndLegality(
```

Comment:
This is very nice to know! Thanks!

Reply:
Yeah this is a life-saver, I was so glad to have found this 😄 This works for nested loops too.